### DISCUSSION TITLE
Make a new release with the h2 whitespace fix

### DISCUSSION THREAD

**Discussion by dradetsky** [curl/curl#9479] [Ideas]
I hit the issue with the RFC 9113 header whitespace with newsboat (see https://github.com/newsboat/newsboat/pull/2177) using curl 7.85.0 and nghttp2 1.49.0

I verified that I could fix the issue if I installed the latest of both nghttp2 and curl. Presumably this is because the fix to nghttp2 to add an option to be non-strict about whitespace is not included in its latest release (although I have requested that they make one: https://github.com/nghttp2/nghttp2/issues/1796), and the fix to make curl use this option (#9448) is not included in the latest curl release. So presumably there should be a new release.

**Comment by bagder** [comment]
I'm sorry but I do not consider this to be an important enough flaw (in curl) to warrant breaking the release cycle for. And for the moment, there is not even a new nghttp2 release that has the option.

I would recommend downgrading nghttp2 to 1.48.0 or apply an nghttp2 patch until nghttp2 1.50.0 and curl 7.86.0 have been released.

**Reply by dradetsky** [reply]
First, it's not that I don't have a resolution for me. I'm on arch linux and so I simply installed `aur/libnghttp2-git` and `aur/curl-git`, and it works fine.

Second, I'm aware that a new nghttp2 release has not yet been made. I'm trying to convince them to make one now for the same reason. Admittedly, without them having made a release, a new curl release won't have the same benefit, but unless I'm missing something it doesn't matter which order the releases are made in, just that both curl and nghttp2 make a release. That said, if you waited until nghttp2 made the release that would make sense.

Third, I apologize for not assuming a solid project like curl had a documented release procedure and looking it up. Absent that, I was applying the release procedure I would use for my own shitty projects, which is basically "Is the tip of master better than the current release version? Would the average user prefer it? If so, make release." That said, it's not clear what is considered a "critical problem." While I don't wish to create more work for you, and this is definitely not an urgent need, it might be helpful if https://github.com/curl/curl/blob/master/docs/RELEASE-PROCEDURE.md were to link to examples of problems considered sufficiently critical that out-of-cadence release was considered warranted.

As for the original issue, I first observed it in newsboat, which uses libcurl. When checking my rss feeds, I hit this error 12 times in 151 feeds, or 8% of the time. While these are obviously not a random sample of urls, I still consider 8% failure rate to be pretty significant. It obviously does not substantially affect users of cli curl, who can resolve this with a simple `--http1.1`, but for those who use libcurl as part of a compiled program, and will typically not have built that program in a way which exposes similar tweaks to the user (as is the case with newsboat), this is more serious. Since libcurl is fairly widespread, it will be incorporated into many large complicated projects, possibly in programming languages I don't understand, where it will be difficult for me to even know how to fix it at the source level even though I know exactly what the problem is; indeed it took a fair amount of time just to figure out how to do the equivalent of `--http1.1` in newsboat. In my case, I was able to resolve this fairly easily with the aur packages, but for many users (with woefully inferior package managers) this may be much more difficult.

So I don't know if this meets the standards, and obviously it's your call. But I think it's fair to say that it isn't a nonissue.

**Reply by bagder** [reply]
I hear you, this is an issue. I did not say nor imply anything else. Most people will however *not* run into this problem simply because they will not the latest nghttp2 version or they will not use http2 servers with trailing whitespace in headers.

We do curl releases *at least* every 56 days. For us to break the cycle and do a panic release the problem has to be serious, and even when we do we need to give the release preparation a week or two. We also add risk by going out of schedule. I do not think this issue is worth the risk nor that big of a hurt to enough number of users.

Finally and most importantly: this issue can be fixed entirely without changing curl and I think the problem is *owned* by nghttp2 primarily. If you can upgrade curl to fix the issue then you most likely can also downgrade nghttp2 to fix it.

This is my personal opinion.

**Reply by dradetsky** [reply]
I think this is overall fair.

However, I'm still in general unsatisfied with the ambiguity of the conditions for an out-of-cadence release. I should stress that this isn't a problem I think you _have to_ solve, but just my personal feedback.

For example, do you think that a problem is not out-of-cadence-release-serious if it can be solved by downgrading to a nearby release, such as one or two releases prior? If so, I think this is pretty reasonable, but also the kind of thing that's easy enough to add to the release procedure doc. Or maybe you think that's usually the case, but not the case if the problem is sufficiently severe, such as permitting remote code execution instead of just failing to work on some urls. I also think this is reasonable, and easy enough to add as a guideline rather than a hard-and-fast rule.

In general, I assume that the rule in the past has been "badger agrees it is serious enough for an out-of-cadence release," which is totally understandable (it's not like I'm suggesting you always should have formal rules instead of just maintainers' judgment). But in that case you could still link to or describe issues which in the past had motivated an out-of-cadence release. This would allow me to check if this was a sufficiently severe issue according to the historical standards of the project. To be fair, I didn't actually track down the release procedure doc until you gave me your opinion, which is maybe an argument for not bothering to add more to it. On the other hand, I wasn't totally confident you understood the impact until I described it in my last post, and by that time I had read it. Overall, it means a situation where I'm describing the impact of a problem and not really understanding whether that's a big enough problem or not, which makes me feel kind of stupid. Again, this is not a critical issue (it's unlikely that you could solve the issue of me feeling stupid in general, no matter how hard you tried), but worth some feedback imo.

**Reply by bagder** [reply]
I've created #9495 to mention patch releases in the release procedure doc.

**Comment by piru** [comment]
There is no way for libcurl to address this issue: You either need to downgrade to nghttp2 1.48 *or* use nghttp2 1.50 that has the control functionality. This is outside of libcurl's control.

Meanwhile there are two options to address this:
- Downgrade to nghttp2 1.48 until 1.50 is out *or*
- Use a patch to revert the strict header checking in 1.49, such as:
```
--- nghttp2-1.49.0.orig/lib/nghttp2_http.c	2022-08-22 15:07:27.000000000 +0300
+++ nghttp2-1.49.0/lib/nghttp2_http.c	2022-09-12 19:38:23.524283433 +0300
@@ -379,14 +379,14 @@
     if (session->server || frame->hd.type == NGHTTP2_PUSH_PROMISE) {
       rv = nghttp2_check_authority(nv->value->base, nv->value->len);
     } else {
-      rv = nghttp2_check_header_value_rfc9113(nv->value->base, nv->value->len);
+      rv = nghttp2_check_header_value(nv->value->base, nv->value->len);
     }
     break;
   case NGHTTP2_TOKEN__SCHEME:
     rv = check_scheme(nv->value->base, nv->value->len);
     break;
   default:
-    rv = nghttp2_check_header_value_rfc9113(nv->value->base, nv->value->len);
+    rv = nghttp2_check_header_value(nv->value->base, nv->value->len);
   }
 
   if (rv == 0) {
```

**Reply by dradetsky** [reply]
@piru this is not a question of whether & how a user can fix this issue for themselves. As I said, I have already fixed it for me (using Arch AUR packages). It's a question of whether curl should do a special release to address this issue (to which I now agree: probably not).

> There is no way for libcurl to address this issue

If by "address" you mean "fix, independent of what anyone else does, in a way which is actually reasonable in terms of effort" then I agree. However, the unreleased commits do address the issue insofar as they cause curl to fix the issue if curl uses a version of nghttp2 which allows them to fix it. Which was exactly the extent of addressing the issue I was suggesting should be released.

**Reply by piru** [reply]
Fair enough.

Either way, it is unfortunate that nghttp2 dropped the ball like this in the first place.

**Comment by bagder** [comment]
[nghttp2 1.50.0](https://github.com/nghttp2/nghttp2/releases/tag/v1.50.0) has been released now, which includes the new function to disable the trailing whitespace check.

One of the most frequently reported offender of trailing whitespace, Wordpress, has said [they fixed their issue](https://twitter.com/wordpressdotcom/status/1572587600652955649).

**Comment by bagder** [comment]
The [fosdem site](https://www.fosdem.org/) has the same issue.
